src/lib/logs/log_forwarder.py

```diff
@@ -12,134 +12,75 @@
-    LOGS_SUBSCRIPTION_ID, \
-    PROCESSING_WORKERS, PROCESSING_WORKER_PULL_REQUEST_MAX_MESSAGES, REQUEST_BODY_MAX_SIZE
+from lib.logs.log_forwarder_variables import (
+    LOGS_SUBSCRIPTION_ID,
+    LOGS_SUBSCRIPTION_PROJECT,
+    MAX_SFM_MESSAGES_PROCESSED,
+    PROCESSING_WORKERS,
+    LOG_PROCESS_STARTUP_DELAY_SECONDS,
+)
 from lib.logs.log_self_monitoring import create_sfm_loop
-from lib.logs.logs_processor import _prepare_context_and_process_message
-from lib.logs.worker_state import WorkerState
-from lib.utilities import chunks
+from lib.logs.log_integration_service import LogIntegrationService
+
 
+def run_logs_wrapper(logging_context, instance_metadata, process_number):
+    asyncio.run(run_logs(logging_context, instance_metadata, process_number))
 
-def run_logs(logging_context: LoggingContext, instance_metadata: InstanceMetadata):
+
+async def run_logs(
+    logging_context: LoggingContext, instance_metadata: InstanceMetadata, process_number
+):
     if not LOGS_SUBSCRIPTION_PROJECT or not LOGS_SUBSCRIPTION_ID:
         raise Exception(
-            "Cannot start pubsub streaming pull - GCP_PROJECT or LOGS_SUBSCRIPTION_ID are not defined")
+            "Cannot start pubsub pulling - GCP_PROJECT or LOGS_SUBSCRIPTION_ID are not defined"
+        )
+
+    # Each process starts later than the previous one
+    await asyncio.sleep(process_number * LOG_PROCESS_STARTUP_DELAY_SECONDS)
 
     sfm_queue = Queue(MAX_SFM_MESSAGES_PROCESSED)
-    subscriber_client = pubsub.SubscriberClient()
-    subscription_path = subscriber_client.subscription_path(LOGS_SUBSCRIPTION_PROJECT, LOGS_SUBSCRIPTION_ID)
-
-    # Open worker threads to process logs from PubSub queue and ingest them into DT
-    for i in range(0, PROCESSING_WORKERS):
-        threading.Thread(target=pull_and_flush_logs_forever,
-                         args=(f"Worker-{i}", sfm_queue, subscriber_client, subscription_path,),
-                         name=f"worker-{i}").start()
-
-    # Create loop with a timer to gather self monitoring metrics and send them to GCP (if enabled)
-    create_sfm_loop(sfm_queue, logging_context, instance_metadata)
-
-
-def pull_and_flush_logs_forever(worker_name: str,
-                                sfm_queue: Queue,
-                                subscriber_client: SubscriberClient,
-                                subscription_path: str):
-    logging_context = LoggingContext(worker_name)
-    worker_state = WorkerState(worker_name)
-    pull_request = PullRequest()
-    pull_request.max_messages = PROCESSING_WORKER_PULL_REQUEST_MAX_MESSAGES
-    pull_request.subscription = subscription_path
+    log_integration_service = LogIntegrationService(sfm_queue)
+    await log_integration_service.update_gcp_client(logging_context)
+
+    tasks = []
+
+    update_token = asyncio.create_task(log_integration_service.keep_gcp_token_updated(logging_context))
+
+    tasks.append(update_token)
+    for worker_number in range(0, PROCESSING_WORKERS):
```

Comment:
So, I got confused. Do we still need this for? I believe based on tests, yes.
But we have individual sets of coroutines for pulling and pushing, this on top of it, and multiprocessing on top of everything??